### PR TITLE
[BUGFIX] TSDB: Fix race on stale values in headAppender

### PR DESCRIPTION
**Pull request by bboreham**

### CODE REVIEW

File: tsdb/head_append.go [hunk 1]
```diff
@@ -357,20 +357,22 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
 	}
 
 	if value.IsStaleNaN(v) {
-		// This is not thread safe as we should be holding the lock for "s".
+		s.Lock()
```
Comment (machine424) [inline tsdb/head_append.go:360]: nit: for simplicity and assuming it's ok to run `value.IsStaleNaN(v)` with the code held, we can move this before the `if` (if not, maybe add a comment so it's not moved for that reason in the future)

**Review by machine424 — APPROVED** [review]
good catch, I assume that's why you're trying to enable the race detector back (I was going to take care of that to help with the tests parallelization PRs)

**Comment by beorn7** [comment]
I vaguely remember that @krajorama had a reason to not apply the "obvious" fix. @krajorama could you have a look here and make sure we are not missing anything?

**Comment by krajorama** [comment]
> I vaguely remember that @krajorama had a reason to not apply the "obvious" fix. @krajorama could you have a look here and make sure we are not missing anything?

I didn't have time to figure it out when I did #15142 .

My own vague memory is that I was concerned with a race between series creation `s, _, err = a.getOrCreate(lset)` and this code.

Let me see the scenarios after #15142.
1. Either this goroutine locks the series first and sees empty lastH* and adds a float NaN.
2. An other goroutine arrives first and sets lastH* and here we add a stale histogram.

Both should be handled correctly later, so this seems fine.

File: tsdb/head_append.go (hunk 1 above)
Comment (beorn7) [reply tsdb/head_append.go:360]: It looks indeed that we can move the `s.Lock` outside of the `if` (and remove the `else` branch below).

Maybe I'm missing the same thing as @machine424 , but then it definitely warrants a comment.

**Review by beorn7 — COMMENTED** [review]
Thanks for your response @krajorama .

File: tsdb/head_append.go (hunk 1 above)
Comment (bboreham) [reply tsdb/head_append.go]: Agreed; I posted the simplification.  Please squash if merging.

**Pull request merged**